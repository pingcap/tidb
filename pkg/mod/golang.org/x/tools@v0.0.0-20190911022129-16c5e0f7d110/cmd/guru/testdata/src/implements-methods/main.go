package main

// Tests of 'implements' query applied to methods.
// See go.tools/guru/guru_test.go for explanation.
// See implements-methods.golden for expected query results.

import _ "lib"

func main() {
}

type F interface {
	f() // @implements F.f "f"
}

type FG interface {
	f()       // @implements FG.f "f"
	g() []int // @implements FG.g "g"
}

type C int
type D struct{}

func (c *C) f() {} // @implements *C.f "f"
func (d D) f()  {} // @implements D.f "f"

func (d *D) g() []int { return nil } // @implements *D.g "g"

type sorter []int

func (sorter) Len() int           { return 0 } // @implements Len "Len"
func (sorter) Less(i, j int) bool { return false }
func (sorter) Swap(i, j int)      {}

type I interface {
	Method(*int) *int // @implements I.Method "Method"
}
