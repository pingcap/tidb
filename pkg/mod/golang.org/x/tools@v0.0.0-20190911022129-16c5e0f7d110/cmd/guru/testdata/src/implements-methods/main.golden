-------- @implements F.f --------
abstract method func (F).f()
	is implemented by method (*C).f
	is implemented by method (D).f
	is implemented by method (FG).f

-------- @implements FG.f --------
abstract method func (FG).f()
	is implemented by method (*D).f
	implements method (F).f

-------- @implements FG.g --------
abstract method func (FG).g() []int
	is implemented by method (*D).g

-------- @implements *C.f --------
concrete method func (*C).f()
	implements method (F).f

-------- @implements D.f --------
concrete method func (D).f()
	implements method (F).f
concrete method func (D).f()
	implements method (FG).f

-------- @implements *D.g --------
concrete method func (*D).g() []int
	implements method (FG).g

-------- @implements Len --------
concrete method func (sorter).Len() int
	implements method (lib.Sorter).Len

-------- @implements I.Method --------
abstract method func (I).Method(*int) *int
	is implemented by method (lib.Type).Method

